[WIP] Add a test that lints factories

imtayadeway opened the pull request:
There are 80+ factories that are currently invalid, and it would be
great to get that number down. Adding this task will help with doing
that in small steps.

The `:rr_pending_change` factory causes some issues that break the
linter, so the linter will skip it for now.

Fryguy commented on lib/tasks/factory_girl.rake:
```diff
@@ -0,0 +1,22 @@
+        DatabaseCleaner.clean
+      end
+    else
+      system("bundle exec rake factory_girl:lint RAILS_ENV='test'")
```
You can do this just by setting ENV directly up above.  See how we do it for other tasks.

Fryguy commented:
Can you define "invalid"?

imtayadeway commented:
@Fryguy they are invalid in that they alone don't satisfy the minimum requirements for validation of their model, i.e. calling `FactoryGirl.create(:factory_in_question)` will fail validation. I believe the linter will also reveal issues concerning uniqueness, i.e. if you create several of the same factory one of them will fail on a uniqueness validation. I believe there are other issues....it looks like some of factories no longer have a model (and should be candidates for deletion). Others aren't used and hence could be deleted too.

I don't know how contentious it is to say that your factories should all be valid in the sense outlined above. I do think there is some value in working towards this, though.

chrisarcand commented:
This is always a weird line to draw because on one hand expecting factories to be valid by default makes understanding easier and test code prettier. On the other hand, if you require a lot of associations it can turn into a very large mess. This can be mitigated if you're ok with just setting ID columns that don't correspond to an actual row in the other table, though...

@Fryguy [Here](https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#linting-factories) is the inspiration for this, I imagine.

chrisarcand commented:
FWIW I'm +1 on this assuming we do not add it as a required task in our suite. More just a command to occasionally inform us of the state of our factories. My biggest qualm, actually, is just that we have database_cleaner added to the list of installed gems just for this one little aside. 😕

imtayadeway commented:
@chrisarcand thanks for input! And for correctly linking to the FactoryGirl docs, where this comes from. I'm not dead set on getting this merged, but thought it would be good to weigh in first before (and if) I make any future changes to factories.

IIRC you used to be able to lint factories at the start of your test suite....I wouldn't want to enable this but might be worth investigating if it could lean on the test framework (and hence eliminate the need for database cleaner in the rake task).

imtayadeway commented:
@miq-bot add-label wip

imtayadeway commented:
I removed that task and added this as a test instead. The test will only run with the `--tag` option set to `lint`. Making this a test avoids having to depend on a new gem (database cleaner) to clean up.

Putting this in the factories directory feels weird, open to suggestions there.

/cc @chrisarcand @Fryguy

chrisarcand commented:
¯\\\_(ツ)_/¯ Certainly addresses the db cleaner concern I had. I'm ok with this I guess.
